### PR TITLE
allow upload of multi-series charms

### PR DESCRIPTION
**Pull request by mhilton**
This is mostly a set of tests for uploading multi-series charms through the v4 interface.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1252/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1255/
Test PASSed.

**Comment by mhilton** [comment]
## QA Instructions

This is a little trickier than it could be as the client isn't ready yet.

Here is a little script that will help you upload charm directories.

``` sh
#!/bin/sh
# testUploadArchive <dir> <id>

URL=${JUJU_CHARMSTORE:-http://localhost:8080}
OLDDIR=`pwd`
cd $1
TEMPDIR=`mktemp -d`
zip -q -r $TEMPDIR/archive.zip *
HASH=`sha384sum $TEMPDIR/archive.zip | awk '{print $1}'`
bhttp POST "${URL}/v4/$2/archive?hash=$HASH" 'Content-Type:application/zip' --stdin < $TEMPDIR/archive.zip
rm -rf $TEMPDIR
cd $OLDDIR
```

it requires @rogpeppe's bhttp command `go get github.com/rogpeppe/bhttp`

Save that to testUploadArchive and set it executable then you can upload charms.

To QA try uploading different charms to different names and see what happens. To create a mutli-series charm add a series parameter to metadata.yaml containing a list of series.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1260/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1261/
Test PASSed.

**Comment by urosj** [comment]
:+1: QA ok.

File: internal/charmstore/store.go [hunk 1]
```diff
@@ -526,9 +526,16 @@ func (s *Store) AddCharm(c charm.Charm, p AddParams) (err error) {
 	if p.URL.URL.Series == "bundle" || p.URL.URL.User == "" || p.URL.URL.Revision == -1 {
```
Comment (rogpeppe) [inline internal/charmstore/store.go]: how about:

```
id := &p.URL.URL
```

then you can get rid of many occurrences of p.URL.URL below ?

File: internal/charmstore/store.go [hunk 2]
```diff
@@ -526,9 +526,16 @@ func (s *Store) AddCharm(c charm.Charm, p AddParams) (err error) {
...
 		return errgo.Newf("charm added with invalid id %v", &p.URL.URL)
 	}
-	if p.URL.URL.Series == "" && len(c.Meta().Series) == 0 {
-		return errgo.Newf("charm added without series %v", &p.URL.URL)
+	if len(c.Meta().Series) == 0 {
```
Comment (rogpeppe) [inline internal/charmstore/store.go]: I'm not sure we should allow putting a multi-series charm to an id with a series.

So wouldn't this be better as:

```
 if id.Series == "" && len(c.Meta().Series) == 0 {
      return errgo.WithCausef(nil, params.ErrBadRequest, "charm %v added without any supported series", &p.URL.URL)
 }
```

Then you don't need the contains logic either.

File: internal/charmstore/store.go (hunk 1 above)
Comment (mhilton) [inline internal/charmstore/store.go]: done

File: internal/charmstore/store.go [hunk 3]
```diff
@@ -526,9 +526,16 @@ func (s *Store) AddCharm(c charm.Charm, p AddParams) (err error) {
...
+		if p.URL.URL.Series == "" {
+			return errgo.WithCausef(nil, params.ErrBadRequest, "charm added without series %v", &p.URL.URL)
+		}
+	} else {
```
Comment (rogpeppe) [inline internal/charmstore/store.go]: s/else {/else if {/

?

File: internal/charmstore/store.go [hunk 4]
```diff
@@ -526,9 +526,16 @@ func (s *Store) AddCharm(c charm.Charm, p AddParams) (err error) {
 	if p.URL.URL.Series == "bundle" || p.URL.URL.User == "" || p.URL.URL.Revision == -1 {
 		return errgo.Newf("charm added with invalid id %v", &p.URL.URL)
 	}
-	if p.URL.URL.Series == "" && len(c.Meta().Series) == 0 {
-		return errgo.Newf("charm added without series %v", &p.URL.URL)
+	if len(c.Meta().Series) == 0 {
+		if p.URL.URL.Series == "" {
+			return errgo.WithCausef(nil, params.ErrBadRequest, "charm added without series %v", &p.URL.URL)
```
Comment (rogpeppe) [inline internal/charmstore/store.go]: I'm not entirely sure that a Store method should be returning ErrBadRequest. This method doesn't necessarily know that it's being called directly as a result of a request.

Comment (rogpeppe) [inline internal/charmstore/store.go]: At the least, this method should document the error causes that it returns.

File: internal/charmstore/store.go [hunk 5]
```diff
@@ -561,7 +568,7 @@ func (s *Store) AddCharm(c charm.Charm, p AddParams) (err error) {
 			return errgo.Newf("charm name duplicates bundle name %v", entity.URL)
```
Comment (rogpeppe) [inline internal/charmstore/store.go]: I wonder if this isn't a "duplicate upload" too.

File: internal/charmstore/store.go (hunk 2 above)
Comment (rogpeppe) [inline internal/charmstore/store.go]: I think my initial confusion arose from the fact we're switching on the meta series, not the uploaded series.

An alternative (hopefully equivalent) phrasing, using the uploaded series to drive the logic, which fits better with the way my brain sees it:

```
if id.Series == "" {
     if len(c.Meta().Series) == 0 {
         return errgo.WithCausef(nil, params.ErrBadRequest, "charm %v added without any supported series", &p.URL.URL)
     }
 } else if len(c.Meta.Series) != 0 && !contains(id.Series, c.Meta().Series) {
     return errgo.WithCausef(nil, params.ErrForbidden, "%s not listed in charm metadata", p.URL.URL.Series)
 }
```

Take it or leave it.

**Comment by rogpeppe** [comment]
LGTM with a few minor suggestions, thanks!

File: internal/charmstore/store.go (hunk 2 above)
Comment (mhilton) [inline internal/charmstore/store.go]: I like that, done.

File: internal/charmstore/store.go (hunk 4 above)
Comment (mhilton) [inline internal/charmstore/store.go]: I'll defer this to a later PR

File: internal/charmstore/store.go (hunk 5 above)
Comment (mhilton) [inline internal/charmstore/store.go]: I think that's reasonable, done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1266/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1267/
Test PASSed.

**Comment by mhilton** [comment]
:shipit: thanks for the reviews

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**